src/components/organisms/NavBarSchedule/utils.ts

```diff
@@ -0,0 +1,38 @@
+import {
+  differenceInMinutes,
+  endOfDay,
+  getUnixTime,
+  max,
+  min,
+  startOfDay,
+} from "date-fns";
+
+import { PersonalizedVenueEvent, VenueEvent } from "types/venues";
+import { MyPersonalizedSchedule } from "types/User";
+
+import { WithVenueId } from "utils/id";
+import { eventEndTime, eventStartTime } from "utils/event";
+import { isTruthy } from "utils/types";
+
+export const prepareForSchedule = (
+  day: Date,
+  usersEvents: MyPersonalizedSchedule
+) => (event: WithVenueId<VenueEvent>): PersonalizedVenueEvent => {
+  const startOfEventToShow = max([eventStartTime(event), startOfDay(day)]);
+  const endOfEventToShow = min([eventEndTime(event), endOfDay(day)]);
+
+  return {
+    ...event,
+    start_utc_seconds: getUnixTime(startOfEventToShow),
+    duration_minutes: differenceInMinutes(endOfEventToShow, startOfEventToShow),
+    isSaved: isTruthy(
+      event.id && usersEvents[event.venueId]?.includes(event.id)
+    ),
+  };
+};
+
+export const buildLocationString = (event: WithVenueId<VenueEvent>) =>
+  `${event.venueId}#${event.room ?? ""}`;
+
+export const extractLocation = (locationStr: string) =>
+  locationStr.split("#", 2);
```

Comment:
I needed here a key to group the events. I consider group unique by venueId and roomTitle.  For that I create a string
"venueId + "#" + "roomTitle" (`buildLocationString`) and then parse it back (`extractLocation`). I believe I could go without `extractLocation` and just to take venueId and roomTitle from the 1st event in the group. But thought, that creating `extractLocation` is a clealer solution.

<hr />

By the way, I believe it's worth mentioning in the world-architecture related issues (if there is no such information yet) that it would be better to have global room IDs / event IDs to connect them properly. Because currently if there are 2 rooms within one venue with the same name, it'll be uncertain to which exactly room the event is related to (as event stores only room name).

Reply:
> I needed here a key to group the events. I consider group unique by venueId and roomTitle.

That makes sense. I think it looked weird to me since we already have the somewhat similar `getExternalRoomSlug` in `src/utils/room.ts`; which has been used to create these venue/room pairs in the past.

I'm not 100% sure if `venueName` and `roomTitle` are guaranteed not to contain `/` though, so unsure if this would work to 'reverse the id' as you need to.

---

> By the way, I believe it's worth mentioning in the world-architecture related issues (if there is no such information yet) that it would be better to have global room IDs / event IDs to connect them properly. Because currently if there are 2 rooms within one venue with the same name, it'll be uncertain to which exactly room the event is related to (as event stores only room name).

Yup, definitely agree.

I mentioned this on slack the other day in https://doesitsparkle.slack.com/archives/C01ER4M61TL/p1621551150069900 (after seeing this comment via email notification); but to summarise here:

> For the room part specifically, we have this backlog issue for re-architecting that: https://github.com/sparkletown/internal-sparkle-issues/issues/133 (which is sort of a semi-related/adjacent plan alongside the ‘world architecture’ refactors: https://github.com/sparkletown/internal-sparkle-issues/issues/127)